File: Solutions/135-candy/candy.php
class Solution {

    /**
     * @param Integer[] $ratings
     * @return Integer
     */
    function candy($ratings) {
        $n = count($ratings);
        
        // Edge case: if there's only one child, give 1 candy
        if ($n == 1) return 1;
        
        // Initialize candies array where each child gets at least 1 candy
        $candies = array_fill(0, $n, 1);
        
        // Step 1: Traverse from left to right
        // If the current child has a higher rating than the previous child,
        // then current child must have more candies than the previous child
        for ($i = 1; $i < $n; $i++) {
            if ($ratings[$i] > $ratings[$i - 1]) {
                $candies[$i] = $candies[$i - 1] + 1;
            }
        }
        
        // Step 2: Traverse from right to left
        // If the current child has a higher rating than the next child,
        // then current child must have more candies than the next child
        // But also, it must maintain the max between what it currently has and this condition
        for ($i = $n - 2; $i >= 0; $i--) {
            if ($ratings[$i] > $ratings[$i + 1]) {
                $candies[$i] = max($candies[$i], $candies[$i + 1] + 1);
            }
        }
        
        // Step 3: Sum up all the candies to get the minimum required candies
        $result = 0;
        for ($i = 0; $i < $n; $i++) {
            $result += $candies[$i];
        }
        
        return $result;
    }
}

File: Solutions/135-candy/candy.rb
# @param {Integer[]} ratings
# @return {Integer}
def candy(ratings)
  n = ratings.length
  candies = Array.new(n, 1)  # Every child gets at least 1 candy

  # Pass 1: left to right
  # If the current child has a higher rating than the previous child,
  # give them one more candy than the previous child
  (1...n).each do |i|
    if ratings[i] > ratings[i - 1]
      candies[i] = candies[i - 1] + 1
    end
  end

  # Pass 2: right to left
  # If the current child has a higher rating than the next child,
  # ensure they have more candies than the next child
  (n - 2).downto(0) do |i|
    if ratings[i] > ratings[i + 1]
      candies[i] = [candies[i], candies[i + 1] + 1].max
    end
  end

  # The minimum number of candies is the sum of all candies
  candies.sum
end
